### PR TITLE
DEP,BUG: Coercion/cast of array to a subarray dtype will be fixed

### LINKED ISSUE #17511: Initializing arrays with subarray dtype
#17419, which deprecated the use of structured arrays composed of just a subarray, is causing a few problems in astropy (see https://github.com/astropy/astropy/issues/10815). While these can be fixed with the suggestion made by @seberg, one of the fixes is not very nice: In our case, it turns out all problems are with `np.array([], dtype=subarray)`, which means we have to special-case an empty list.

But I wonder if it doesn't make sense to try to move towards the case where
```
np.array(data, dtype)
```
always returns an array with the requested dtype, and for `data=[]` it would be equivalent to `np.empty(0, dtype)`. Certainly, it is strange that the first of the following works but the second does not:
```
>>> import numpy as np
>>> np.array(([1, 2], 3), dtype='(2,)i,i')
array(([1, 2], 3), dtype=[('f0', '<i4', (2,)), ('f1', '<i4')])
>>> np.array(([1, 2],), dtype='(2,)i')
TypeError: int() argument must be a string, a bytes-like object or a number, not 'tuple'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: setting an array element with a sequence.
```

Of course, I realize it is not easy to move from arguably mistaken behaviour to more logical one, but looking at our problem cases at least, I realize that in those we basically assumed the more logical one (i.e., they were bugs that somehow failed to get exposed)...

Concretely, I wonder if one could warn and move to more logical behaviour in one go (or give the option to use the more logical behaviour or so).

### PR DESCRIPTION
**Pull request by seberg**
This currently appends the subarray dtype dimensions first and then
tries to assign to the result array which uses incorrect broadcasting
(broadcasting against the subarray dimensions instead of repeating
each element according to the subarray dimensions).

This also fixes the python scalar pathway `np.array(2, dtype="(2)f4,")`
which previously only filled the first value.  I consider that a clear
bug fix.

Closes gh-17511

### CODE REVIEW

File: numpy/core/tests/test_dtype.py [hunk 1]
```diff
@@ -314,6 +314,23 @@ def test_union_struct(self):
                        'formats':['i1', 'O'],
                        'offsets':[np.dtype('intp').itemsize, 0]})
 
+    @pytest.mark.parametrize(["obj", "dtype", "expected"],
+        [(3, "(3)f4,", [3, 3, 3]),
+         (np.float64(2), "(2)f4,", [2, 2]),
```
Comment (seberg) [inline numpy/core/tests/test_dtype.py:319]: To be clear, the first case here previously returned `[3, uninitialized, uninitialized]`, while the second case actually did the right thing.  So I considered that a simple bugfix here.

File: doc/release/upcoming_changes/17596.future.rst [hunk 2]
```diff
@@ -0,0 +1,30 @@
+Arrays cannot be using subarray dtypes
+--------------------------------------
```
Comment (mattip) [inline doc/release/upcoming_changes/17596.future.rst:2]: Not sure about the length of the underline, pl
```suggestion
Arrays cannot be using subarray dtypes
Clean up inconsistencies in using subarray dtypes
-------------------------------------------------
```

**Comment by mhvk** [comment]
@seberg - this looks quite good. With this change, will `np.array([], '2f4')` still raise a `FutureWarning`? I think it doesn't have to, since the result will not change in the future, but I didn't see a test to check that (and didn't quite follow the code close enough to see if this special case was caught).

**Comment by seberg** [comment]
Thanks for taking a look! I am scanning the full input for array(-like) objects in the code, it is a bit annoying, but feel this is fringe enough that it is OK.  Since I look for array likes, it will not trigger for empty sequences.

There are cases where a warning will be given even though nothing will change, e.g. `np.array(3).astype("(4)f4")` has the same future behaviour as with the current broadcast behaviour.

**Comment by mhvk** [comment]
OK, that sounds good! And I can see that you cannot do everything...  But perhaps good to add a test for the empty list case, just in case?

**Comment by seberg** [comment]
Oh, seems I accidentally deleted that sentence snippet: I had just added the test.

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**